File: LPIS_1/src/main/java/com/example/lab/interpreter/InterpreterContextImpl.java
package com.example.lab.interpreter;

import com.example.lab.FunctionSignature;
import com.example.lab.ReturnType;
import com.example.lab.Type;
import com.example.lab.statement.StatementContext;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class InterpreterContextImpl implements InterpreterContext {

    private final StatementContext context;
    private final Map<String, Object> values = new HashMap<>();

    public InterpreterContextImpl(StatementContext context) {
        this.context = context;
    }

    @Override
    public void returnVoid() {
    }

    @Override
    public void callFunctionVoid(String name, List<Object> arguments) {
        if (name.equals("print")) {
            System.out.println(arguments.get(0));
        }
    }

    @Override
    public void setValue(String name, int value) {
        values.put(name, value);
    }

    @Override
    public void setValue(String name, List<Integer> value) {
        values.put(name, value);
    }

    @Override
    public int getValueElement(String name) {
        return (Integer) values.get(name);
    }

    @Override
    public List<Integer> getValueList(String name) {
        return (List<Integer>) values.get(name);
    }

    @Override
    public InterpreterContext block() {
        var block = new InterpreterContextImpl(context.block());
        block.values.putAll(values);
        return block;
    }

    @Override
    public Optional<ReturnType> getReturnType(FunctionSignature function) {
        return context.getReturnType(function);
    }

    @Override
    public Optional<Type> getVariableType(String name) {
        return context.getVariableType(name);
    }

}

File: LPIS_1/src/main/java/com/example/lab/interpreter/InterpreterContextImpl.kt
package com.example.lab.interpreter

import com.example.lab.FunctionSignature
import com.example.lab.expression.Expression
import com.example.lab.statement.StatementContext

class InterpreterContextImpl(private val context: StatementContext) : InterpreterContext {

	private val values: MutableMap<String?, Any> = HashMap()

	override fun returnVoid() {
	}

	override fun callFunctionVoid(name: String, arguments: List<Any>) {
		if(name == "print") {
			println(arguments[0])
		}
	}

	override fun setValue(name: String, value: Int) {
		values[name] = value
	}

	override fun setValue(name: String?, value: List<Int>) {
		values[name] = value
	}

	override fun getValueElement(name: String): Int {
		return values[name] as Int
	}

	override fun getValueList(name: String): List<Int> {
		return values[name] as List<Int>
	}

	override fun block(): InterpreterContext {
		val block = InterpreterContextImpl(context.block())
		block.values.putAll(values)
		return block
	}

	override fun getReturnTypeOrNull(function: FunctionSignature): Expression.Type? {
		return context.getReturnTypeOrNull(function)
	}

	override fun getVariableTypeOrNull(name: String): Expression.Type? {
		return context.getVariableTypeOrNull(name)
	}
}
